R/DataModel.R

```diff
@@ -310,73 +322,100 @@ naToZero <- function(x) {
+      on.exit(DatabaseConnector::disconnect(connection))
+    } else {
+      stop("No connection or connectionDetails provided.")
+    }
   }
 
-  uploadTable <- function(tableName) {
-    ParallelLogger::logInfo("Uploading table ", tableName)
+  if (connection@dbms == "sqlite" & schema != "main") {
+    stop("Invalid schema for sqlite, use schema = 'main'")
+  }
 
-    primaryKey <- specifications %>%
-      dplyr::filter(tableName == !!tableName &
-                      tolower(primaryKey) == "yes") %>%
-      dplyr::select("fieldName") %>%
-      dplyr::pull()
+  start <- Sys.time()
 
-    if (purgeSiteDataBeforeUploading &&
-      "database_id" %in% primaryKey) {
+  # TODO: Seems specific to CD? What if the cdmSourceFile does not exist?
+  # I think the logic for obtaining the databaseId should be in the uploadTable
+  # function.
+  # if (purgeSiteDataBeforeUploading) {
+  #   database <-
+  #     readr::read_csv(file = file.path(unzipFolder, cdmSourceFile),
+  #                     col_types = readr::cols())
+  #   colnames(database) <-
+  #     SqlRender::snakeCaseToCamelCase(colnames(database))
+  #   databaseId <- database$databaseId
+  # }
 
-      type <- specifications %>%
+  uploadTable <- function(tableName) {
+    csvFileName <- paste0(tableName, ".csv")
+    if (csvFileName %in% list.files(resultsFolder)) {
+      rlang::inform(paste0("Uploading file: ", csvFileName, " to table: ", tableName))
+
+      primaryKey <- specifications %>%
         dplyr::filter(tableName == !!tableName &
-                        fieldName == "database_id") %>%
-        dplyr::select("dataType") %>%
+                        tolower(primaryKey) == "yes") %>%
+        dplyr::select("columnName") %>%
         dplyr::pull()
 
-      deleteAllRowsForDatabaseId(
-        connection = connection,
-        schema = schema,
-        tableName = paste0(tablePrefix, tableName),
-        databaseId = databaseId,
-        idIsInt = type %in% c("int", "bigint")
-      )
-    }
-
-    csvFileName <- paste0(tableName, ".csv")
-    if (csvFileName %in% list.files(unzipFolder)) {
+      # Create an environment variable to hold
+      # the information about the target table for
+      # uploading the data
       env <- new.env()
       env$schema <- schema
       env$tableName <- paste0(tablePrefix, tableName)
       env$specTableName <- tableName
       env$primaryKey <- primaryKey
       if (purgeSiteDataBeforeUploading &&
         "database_id" %in% primaryKey) {
+        # Get the databaseId by reading the 1st row of data
+        results <- readr::read_csv(
+          file = file.path(resultsFolder, csvFileName),
+          n_max = 1,
+          show_col_types = FALSE)
+
+        type <- specifications %>%
+          dplyr::filter(tableName == !!tableName &
+                          columnName == "database_id") %>%
+          dplyr::select("dataType") %>%
+          dplyr::pull()
+
+        # TODO: The databaseId may be missing if there are
```

Comment:
Related to the comment above - how we obtain the databaseId is important when we are attempting to purge prior results. The logic I have provided here will fail if: results were uploaded to a table previously, then a new run produces 0 results and we'd like to purge the old results...then the databaseId will not be present in the new data and as a result, the old results will not be purged.